### ISSUE TITLE
builder.save() has strange behavior (it might be a bug as described in the second update)

### ISSUE THREAD

**Issue by atomextranova** [tensorflow/serving#529]
Hi all, I encountered a strange problem when I tried to save my model. When I train my model, I load weights and biases from a pre-trained npy file with codes like:
~~~
        # Load the weights into memory
        weights_dict = np.load(self.WEIGHTS_PATH, encoding='bytes').item()

        # Loop over all layer names stored in the weights dict
        for op_name in weights_dict:

            # Check if the layer is one of the layers that should be reinitialized
            if op_name not in self.SKIP_LAYER:

                with tf.variable_scope(op_name, reuse=True):

                    # Loop over list of weights/biases and assign them to their corresponding tf variable
                    for data in weights_dict[op_name]:

                        # Biases
                        if len(data.shape) == 1:

                            var = tf.get_variable('biases', trainable=False)
                            session.run(var.assign(data))

                        # Weights
                        else:

                            var = tf.get_variable('weights', trainable=False)
                            session.run(var.assign(data))
~~~
And the Saver.saver() gives me the correct files.(.meta file is about 10mb)
However when I again used the simillar codes with builder.save(), the pb file (which should only contain the graph I save, I think) is larger than 1GB (which is not allowed by default server option and apparently has all variable values stored in .00000-of-00001 file). I suppose it is builder.add_meta_graph_and_variable() mistakenly saved the temporary variables I got from tf.get_variable() from my graph. I tried builder.add_meta_graph but it said add_meta_graph_and_variabes must be called first. Is there a way not to save these variables in pb files? So could I simply turn off the SetMessageLimit() in main.cc? Codes shown below:

~~~
        reader = tf.train.NewCheckpointReader(FLAGS.file_name)
        temp_weights = reader.get_tensor(current_op_name + '/weights')
        temp_biases = reader.get_tensor(current_op_name + '/biases')
        for index in range(1, 3):
            current_scope = current_op_name + '_%d' % index
            with tf.variable_scope(unicode(current_scope), reuse=True):
                variable_weights = tf.get_variable('weights', trainable=False)
                variable_biases = tf.get_variable('biases', trainable=False)
                sess.run(variable_weights.assign(temp_weights))
                sess.run(variable_biases.assign(temp_biases))
       .....
               legacy_init_op = tf.group(tf.tables_initializer(), name='legacy_init_op')
        builder.add_meta_graph_and_variables(
            sess, [tag_constants.SERVING],
            signature_def_map={
                'predict_images':
                    prediction_signature
            },
            legacy_init_op=legacy_init_op)
        builder.save()
~~~

update1:
I found one way to reduce the size of pb files. I could create the net and use builder.add_meta_graph_and_variables() first, and then load my own weights and biases. This seems to give me the set of files that is of correct size, but the result is wrong vary every time I deploy it with tensorflow_model_server. (For some reason I have to load weights from a network with different graph so I could not use load_from_checkpoint directly.) Also, if I use import_meta_graph() combined with saver.restore(), the model also seems to vary every time when deployed.

update2:
By the way, I could confirm at least one strange feature or bug: if your model has a dropout layer that is trained with keep_prob < 1, and you export it from a checkpoint file with import_meta_graph, the result will vary every time you send the same request!

update3:
The only workable solution to me (which does not satisfy my needs) is that you train a model and save a checkpoint file. And then recreate the same model and use saver.restore, and then save it. This one is really limited and causes coupling between model training and exporting.

**Comment by atomextranova** [comment]
I found one way to reduce the size of pb files. I could create the net and use builder.add_meta_graph_and_variables() first, and then load my own weights and biases. This seems to give me the set of files that is of correct size, but the result is wrong vary every time I use builder.save() from the same check point. (For some reason I have to load weights from a network with different graph so I could not use load_from_checkpoint directly.) Also, if I use import_meta_graph() combined with get_tensor_by_name, the model also seems to vary every time I save it.

**Comment by atomextranova** [comment]
By the way, I could confirm at least one strange feature or bug: if your model has a dropout layer that is trained with keep_prob < 1, and you export it from a checkpoint file with import_meta_graph, the result will vary every time you send the same request!

**Comment by atomextranova** [comment]
The only workable solution to me (which does not satisfy my needs) is that you train a model and save a checkpoint file. And then recreate it and use load_from_checkpoint, and then save it. This one is really limited and causes coupling between model training and export in industry.

**Comment by atomextranova** [comment]
This is really annoying that most of my problems might be caused by the following codes:
~~~
builder.add_meta_graph_and_variables(
      sess, [tag_constants.SERVING],
      signature_def_map={
           'predict_images':
               prediction_signature,
           signature_constants.DEFAULT_SERVING_SIGNATURE_DEF_KEY:
               classification_signature,
      },
      legacy_init_op=legacy_init_op) // This legacy_init_op should be deleted
~~~

I will keep updating this thread for a full description after a full test later. But the legacy_init_op is causing the result to be random in many cases. For instance, when you use import_meta_graph to  rebuild the network. I strongly suggests to delete this line in tutorial because it seems to have no use in the tutorial and actually causes problems in practice.

**Comment by gautamvasudevan** [comment]
Closing due to staleness. If this is still an issue, please file a new updated issue with current steps to reproduce the bug. If this is a question, please ask it on: 

https://stackoverflow.com/questions/tagged/tensorflow-serving

Thanks!
